msgraph/client.go

```diff
@@ -161,100 +149,73 @@ func (c Client) performRequest(req *http.Request, input HttpRequestInput) (*http
+			if f != nil && f(resp, o) {
+				return true, nil
+			}
 		}
+		return retryablehttp.DefaultRetryPolicy(ctx, resp, err)
+	}
 
-		req.Body = io.NopCloser(bytes.NewBuffer(reqBody))
+	req.Body = io.NopCloser(bytes.NewBuffer(reqBody))
 
-		for _, m := range c.RequestMiddlewares {
+	if c.RequestMiddlewares != nil {
+		for _, m := range *c.RequestMiddlewares {
 			r, err := m(req)
 			if err != nil {
 				return nil, status, nil, err
 			}
 			req = r
 		}
+	}
 
-		resp, err = c.httpClient.Do(req)
-		if err != nil {
-			return nil, status, nil, err
-		}
+	resp, err = c.HttpClient.Do(req)
+	if err != nil {
+		return nil, status, nil, err
+	}
 
-		for _, m := range c.ResponseMiddlewares {
+	if c.ResponseMiddlewares != nil {
+		for _, m := range *c.ResponseMiddlewares {
 			r, err := m(req, resp)
 			if err != nil {
 				return nil, status, nil, err
 			}
 			resp = r
 		}
+	}
 
-		o, err = odata.FromResponse(resp)
-		if err != nil {
-			return nil, status, o, err
-		}
+	o, err = odata.FromResponse(resp)
+	if err != nil {
+		return nil, status, o, err
+	}
 
-		if !c.DisableRetries {
-			f := input.GetConsistencyFailureFunc()
-			if f != nil && f(resp, o) {
-				// eventual consistency, just try again
-				if attempts < requestAttemptsForConsistency {
-					continue
-				}
-			}
+	status = resp.StatusCode
+	if !containsStatusCode(input.GetValidStatusCodes(), status) {
+		f := input.GetValidStatusFunc()
+		if f != nil && f(resp, o) {
+			return resp, status, o, nil
 		}
 
-		status = resp.StatusCode
-		if !containsStatusCode(input.GetValidStatusCodes(), status) {
-			f := input.GetValidStatusFunc()
-			if f != nil && f(resp, o) {
-				return resp, status, o, nil
-			}
-
-			rateLimitingStatuses := []int{
-				http.StatusFailedDependency,
-				http.StatusTooManyRequests,
-				http.StatusInternalServerError,
-				http.StatusBadGateway,
-				http.StatusServiceUnavailable,
-			}
-			if containsStatusCode(rateLimitingStatuses, status) {
-				// rate limiting
-				if retryAfter := resp.Header.Get("Retry-After"); retryAfter != "" {
-					if r, err := strconv.ParseFloat(retryAfter, 64); err == nil && r > 0 {
-						// Retry-After header detected, use that instead of exponential backoff
-						backoff = int64(r * float64(time.Second))
-						multiplier = 0
-					}
-				}
-				continue
-			}
-
-			var errText string
-			switch {
-			case o != nil && o.Error != nil && o.Error.String() != "":
-				errText = fmt.Sprintf("OData error: %s", o.Error)
-			default:
-				defer resp.Body.Close()
-				respBody, err := io.ReadAll(resp.Body)
-				if err != nil {
-					return nil, status, o, fmt.Errorf("unexpected status %d, could not read response body", resp.StatusCode)
-				}
-				errText = fmt.Sprintf("response: %s", respBody)
+		var errText string
+		switch {
+		case o != nil && o.Error != nil && o.Error.String() != "":
+			errText = fmt.Sprintf("OData error: %s", o.Error)
+		default:
+			defer resp.Body.Close()
+			respBody, err := io.ReadAll(resp.Body)
```

Comment:
these two lines assume there's a body, but we're not checking that it has any length here?